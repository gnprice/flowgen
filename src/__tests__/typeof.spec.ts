import { compiler, beautify } from "..";
import "../test-matchers";

it("should handle `typeof`", () => {
  const ts = `
export const x: { o: { f(string): number } };
export type Tx = typeof x;
export type Txo = typeof x.o;
export type Txof = typeof x.o.f;

export namespace n { export const x: { y: number } }
export type N = typeof n;
export type Nx = typeof n.x;
export type Nxy = typeof n.x.y;

export namespace n { export namespace nn { export const x: number } }
export namespace n { export module mm { export const x: number } }

// These two should behave exactly the same -- a TS "module" declaration is
// just the old syntax for a TS "namespace", which were originally called
// "internal modules".
export type Nnn = typeof n.nn;
export type Nmm = typeof n.mm;

// Similarly these two.
export type Nnnx = typeof n.nn.x;
export type Nmmx = typeof n.mm.x;
`;
  const result = compiler.compileDefinitionString(ts);
  expect(beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});
