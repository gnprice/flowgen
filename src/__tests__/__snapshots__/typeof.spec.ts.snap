// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle \`typeof\` 1`] = `
"declare export var x: {
  o: {
    f(string: any): number,
    ...
  },
  ...
};
export type Tx = typeof x;
export type Txo = typeof x.o;
export type Txof = typeof x.o.f;
declare var n: typeof npm$namespace$n;

declare var npm$namespace$n: {|
  x: typeof n$x,
  nn: typeof npm$namespace$n$nn,
  mm: typeof npm$namespace$n$mm,
|};
declare export var n$x: {
  y: number,
  ...
};

declare var npm$namespace$n$nn: {|
  x: typeof n$nn$x,
|};
declare export var n$nn$x: number;

declare var npm$namespace$n$mm: {|
  x: typeof n$mm$x,
|};
declare export var n$mm$x: number;
export type N = typeof n;
export type Nx = typeof n.x;
export type Nxy = typeof n.x.y;
export type Nnn = typeof n.nn;
export type Nmm = typeof n.mm;
export type Nnnx = typeof n.nn.x;
export type Nmmx = typeof n.mm.x;
"
`;
